zenoh/src/net/routing/hat/p2p_peer/pubsub.rs

```diff
@@ -277,47 +276,45 @@ pub(super) fn undeclare_client_subscription(
-                if let Some(id) = face_hat_mut!(face).local_subs.remove(res) {
-                    face.primitives.send_declare(RoutingContext::with_expr(
-                        Declare {
-                            interest_id: None,
-                            ext_qos: ext::QoSType::DECLARE,
-                            ext_tstamp: None,
-                            ext_nodeid: ext::NodeIdType::DEFAULT,
-                            body: DeclareBody::UndeclareSubscriber(UndeclareSubscriber {
-                                id,
-                                ext_wire_expr: WireExprType::null(),
-                            }),
-                        },
-                        res.expr(),
-                    ));
-                }
-                for res in face_hat!(face)
-                    .local_subs
-                    .keys()
-                    .cloned()
-                    .collect::<Vec<Arc<Resource>>>()
-                {
-                    if !res.context().matches.iter().any(|m| {
-                        m.upgrade()
-                            .is_some_and(|m| m.context.is_some() && remote_client_subs(&m, face))
-                    }) {
-                        if let Some(id) = face_hat_mut!(&mut face).local_subs.remove(&res) {
-                            face.primitives.send_declare(RoutingContext::with_expr(
-                                Declare {
-                                    interest_id: None,
-                                    ext_qos: ext::QoSType::DECLARE,
-                                    ext_tstamp: None,
-                                    ext_nodeid: ext::NodeIdType::DEFAULT,
-                                    body: DeclareBody::UndeclareSubscriber(UndeclareSubscriber {
-                                        id,
-                                        ext_wire_expr: WireExprType::null(),
-                                    }),
-                                },
-                                res.expr(),
-                            ));
-                        }
+            if let Some(id) = face_hat_mut!(face).local_subs.remove(res) {
+                face.primitives.send_declare(RoutingContext::with_expr(
+                    Declare {
+                        interest_id: None,
+                        ext_qos: ext::QoSType::DECLARE,
+                        ext_tstamp: None,
+                        ext_nodeid: ext::NodeIdType::DEFAULT,
+                        body: DeclareBody::UndeclareSubscriber(UndeclareSubscriber {
+                            id,
+                            ext_wire_expr: WireExprType::null(),
+                        }),
+                    },
+                    res.expr(),
+                ));
+            }
+            for res in face_hat!(face)
+                .local_subs
+                .keys()
+                .cloned()
+                .collect::<Vec<Arc<Resource>>>()
```

Comment:
Is `collect` really needed()

Reply:
Yes. Otherwise there would be a double borrow on tables.